# 우아하게 비동기 처리하기

### 웹서비스에서 가장 다루기 어려운 부분은?

리액트, vue 같은 선언형 프로그래밍이 나오면서 웹이 간편해졌음.

비동기 프로그래밍 ⇒ 순서가 보장되지 않은 상황

다루기가 어려운 이유 ⇒ 우선 좋은 코드에 대한 생각

```jsx
function getBarzFromX(x){
	if(x === undefined){
	  return undefined;
  }
	if(x.foo === undefined){
	  return undefined;
  }	if(x.foo.bar === undefined){
	  return undefined;
  }
	return x.foo.bar.baz
}
```

이것은 좋지 않은 코드이다. 결국 원하는 값은 x.foo.bar.baz인데 하는 일은 단순하나 결국 프로퍼티에 안전하게 접근하길 원함. 

명령어의 노이즈가 많아서 함수가 어떤 역할을 하는지 드러나지 않음

```jsx
function getBazFromX(x){
	return x?.foo?.bar?.baz;
}
```

코드가 간결하다. 함수의 역할을 한 눈에 확인할 수 있다. 성공할 떄 접근하는 표현식과 모양이 큰 차이가 나지 않는다. 같은 역할을 하는 식이 반환할 때랑 비슷하면 좋은 징조

좋은 코드의 특징 ⇒ 성공, 실패의 경우를 분리해 처리할 수 있다. , 비즈니스 로직을 한 눈에 파악할 수 있다. 

어려운 코드의 특징 ⇒ 실패, 성공의 경우가 서로 섞여 처리된다. , 비즈니스 로직을 파악하기 어렵다.

```jsx
function Profile(){
  const foo = useAsyncValue(()=>{
    return fetchFoo();
	});
  
  if(foo.error) return <div>로딩에 실패했습니다.</div>
  if(!foo.data) return <div>로딩 중입니다...</div>
  return <div>{foo.data.name}님 안녕하세요!</div>
}
```

문제점

- 성공하는 경우와 실패하는 경우가 섞여서 처리된다.
- 실패하는 경우에 대한 처리를 외부에 위임하기 어렵다.
- 만약에 비동기가 여러개가 된다면 상황은 더 복잡해진다.

```jsx
function Profile(){
	const foo = useAsyncValue(()=>{
    return fetchFoo();
	});
  const bar = useAsyncValue(()=>{
	if(foo.error || !foo.data){
    return undefined;
  }    
  });
  if(foo.error || bar.error) return <div>로딩에 실패했습니다.</div>

  if(foo.loading || bar.loading) return <div>로딩중입니다.</div>

  return /*foo와 bar로 적합한 처리하기*/
}
```

이렇게 비동기 작업이 2개가 된다면 9가지의 상태가 되고 더욱 코드가 복잡해집니다. 또한 만약에 3개 4개가 된다면 더욱 더 복잡해질 수 있다.

### 데이터 패칭을 위한 Suspense

```jsx
function FooBar(){
  const foo = useAsyncValue(() => fetchFoo());
	const bar = useAsyncValue(() => fetchBar(foo));
  return <div>{foo}{bar}</div>
}
```

Suspense가 목표로 하는 코드는 간단하다.

- 성공한 경우에만 집중할 수 있는
- 로딩 상태와 에러상태가 분리된
- 동기와 거의 같게 사용할 수 있는

```jsx
<ErrorBoundary fallback={<MyErrorPage />}>
	<Suspense fallback={<Loader />}>
	  <FooBar />
	</Suspense>
</ErrorBoundary>
```

try, catch문과 같이 간결하게 코드를 작성할 수 있다.

### 웹서비스의 코드 복잡도를 낮춘 방법

React Hooks는 어떻게 저희 코드 복잡도를 줄였을까요? ⇒ 선언적인 API

비동기 ⇒ suspense로 비동기적인 리소스를 선언하기만 함

함수형 프로그래밍 ⇒ 함수는 필요한 코드 조각을 선언적으로 사용 ⇒ 메모이제이션이 필요하면 useMemo사용

# React Query 에러처리 

### 1. UseEffect

```jsx
import { useEffect } from "react";
import { useQuery } from "react-query";
import { useNavigate } from "react-router-dom";

const fetchCarClasses = () => {
  return fetch("...");
};

const useTestQuery = () => {
  const navi = useNavigate();
  const query = useQuery("getCars", fetchCarClasses, {
    retry: false,
  });
  useEffect(() => {
    if (query.isError) {
      navi("/errorPage");
    }
  }, [query.isError]);

  return query;
};

export default useTestQuery;
```

우선 시작은 에러처리를 할 때, 해당 데이터를 보여주는 UI 컴포넌트에서 하는 것은 복잡함을 유발(비즈니스 로직 분리)할 수 있기 때문에 error처리를 데이터를 받는 커스텀 훅에서 처리하고 싶다라는 생각에서 시작한다. 

에러가 발생하면  `errorPage` 로 이동시키고 싶은데 처음에 선택한 방법은 `useEffect` 이다. 에러가 발생하면 `useEffect`를 이용해서 `isError` 의 상태가 변경되면 호출되게 하고 `isError` 가 true이면 페이지를 이동시킨다.

주의할점

- `retry:false` 를 한 이유: retry는 에러가 발생했을 때 해당 서버에 요청을 다시 보내는 것이다. `default:3` 으로 설정되어 있기 때문에 3번의 요청 후에 `isError:true` 로 변하기 때문에 재요청을 보내지 않는 것으로 설정해주었다.

### 2. onError

```jsx
import { useEffect } from "react";
import { useQuery } from "react-query";
import { useNavigate } from "react-router-dom";

const fetchCarClasses = () => {
  return fetch("...");
};

const useTestQuery = () => {
  const navi = useNavigate();
  const query = useQuery("getCars", fetchCarClasses, {
    retry: false,
    onError: ()=> navi("/errorPage");
  });

  return query;
};

export default useTestQuery;
```

`onError` 를 이용해서 하면 `useEffect`를 사용하는 것보다 간결한 코드로 구현할 수 있다.

# 참고 
참고: https://www.youtube.com/watch?v=FvRtoViujGg

https://tecoble.techcourse.co.kr/post/2021-10-01-react-query-error-handling/
